### ISSUE TITLE
gradient descent의 compute_loss() 질문!

### ISSUE THREAD

**Issue by ijaydoubleu** [toughhyeok/DL-low-code#2]
혹시 **Ax = b** , **x** = [x1; x2] 인 linear system을 풀고 싶었던 거라면 
compute_loss에서 predictions는 Ax 혹은 Ax - b 같은 게 되었어야 하는 건 아닌가요?
행렬을 코딩하려니까 헷갈리네요! ㅎㅎ

```py
def compute_loss(x1, x2, A, b):
    m = len(y)
    x = np.array([[x1, x2]]).T
    predictions = np.dot(A, x)
    return (1/m)*np.sum((predictions - b) ** 2)
```

**Comment by toughhyeok** [comment]
objective function이라고 정의했던  $f(x)=(y-Ax)^{T}(y-Ax)$ 에서 변수 때문에 헷갈릴 수 밖에 없게 작성된 것 같아요 죄송합니당..😢 헷갈리지 않게 다시 쓰면

$$f(\theta)=(y-X^{'}\theta)^{T}(y-X^{'}\theta)$$

여기서 $X^{'}$는 train 독립 변수 data($X$)에 one-vector가 합쳐진 형태 입니다! (vector는 transpose 표시가 되어 있지 않으면 column vector)

$$X^{'}=\begin{bmatrix}X & 1\end{bmatrix}$$

그리고 $\theta$는 우리가 최적화 하고 싶은 모델의 parameter를 뜻합니다!

$$\theta=\begin{bmatrix}a \\
b\end{bmatrix}$$ 

https://github.com/toughhyeok/DL-low-code/blob/toughhyeok/optimizer/gd.py 에 작성된 `compute_loss` 함수의 argument는

```py
def compute_loss(a, b, X, y):
    m = len(y)
    predictions = a * X + b
    return (1 / (2 * m)) * np.sum((predictions - y) ** 2)
```

각각 아래의 의미를 가집니다!

`X`: train 독립 변수 data  (예제에서는 독립 변수가 1개 이니까 vector, 만약 독립변수가 여러 개 라면 matrix)
`a`: 최적화 대상인 coefficient! (scalar)
`b`: 최적화 대상인 intercept! (scalar)
`y`: train 종속 변수 (vector)

그럼 `compute_loss` 함수를 다시 $f(\theta)=(y-X^{'}\theta)^{T}(y-X^{'}\theta)$ 에 맞게 다시 쓴다면

```py
def compute_loss(a, b, X, y):
    theta = np.array([[a], [b]])
    m = len(y) # X.shape[0]
    
    one_vector = np.ones((m, 1))
    X_prime = np.hstack((X, one_vector))

    return 1/(m)*np.sum((y-X_prime.dot(theta))**2)
```

이렇게 쓸 수 있을 것 같아요!

**Comment by toughhyeok** [comment]
`compute_loss`함수 수정하기!

**Comment by ijaydoubleu** [comment]
오 답이 엄청 빠르다 ㅎㅎ 감사해요!

**Comment by boram-jeong** [comment]
오!! 회사에서도 볼 수 있어서 편하네요!! ㅎㅎ
